File: android-app/app/src/main/java/com/proactivesafety/app/RiskEngine.java
package com.proactivesafety.app;

import android.location.Location;

import java.util.Arrays;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

final class RiskEngine {
    private static final List<RiskZone> ZONES = Arrays.asList(
            new RiskZone(
                    "RZ001",
                    "Highway Cross Junction A",
                    10.0154,
                    76.3419,
                    "HIGH",
                    "Four-road highway crossing",
                    12,
                    30,
                    "Poor night visibility"
            ),
            new RiskZone(
                    "RZ002",
                    "Market Road Blind Turn",
                    10.0188,
                    76.3491,
                    "MEDIUM",
                    "Urban blind turn",
                    5,
                    25,
                    "Blocked side view"
            )
    );

    private RiskEngine() {
    }

    static SafetyAssessment assess(Location location) {
        SafetyAssessment assessment = new SafetyAssessment();
        assessment.nearestZone = nearestZone(location);
        assessment.distanceMeters = distanceMeters(location, assessment.nearestZone);

        int score = 0;
        if (assessment.nearestZone != null) {
            RiskZone zone = assessment.nearestZone;

            if (assessment.distanceMeters <= 120) {
                score += 35;
                assessment.factors.add("Inside immediate danger radius");
            } else if (assessment.distanceMeters <= 300) {
                score += 25;
                assessment.factors.add("Approaching known risk zone");
            } else if (assessment.distanceMeters <= 700) {
                score += 12;
                assessment.factors.add("Risk zone ahead");
            }

            if ("CRITICAL".equals(zone.zoneRisk)) {
                score += 28;
                assessment.factors.add("Zone marked CRITICAL");
            } else if ("HIGH".equals(zone.zoneRisk)) {
                score += 22;
                assessment.factors.add("Zone marked HIGH");
            } else if ("MEDIUM".equals(zone.zoneRisk)) {
                score += 12;
                assessment.factors.add("Zone marked MEDIUM");
            }

            if (zone.roadType.toLowerCase(Locale.US).contains("cross")) {
                score += 14;
                assessment.factors.add("Cross-junction conflict points");
            }

            int speedKmh = speedKmh(location);
            if (speedKmh > zone.recommendedSpeedKmh + 20) {
                score += 18;
                assessment.factors.add("Speed is far above recommended limit");
            } else if (speedKmh > zone.recommendedSpeedKmh) {
                score += 10;
                assessment.factors.add("Speed is above recommended limit");
            }

            if (zone.accidentCount >= 10) {
                score += 10;
                assessment.factors.add("High accident history");
            } else if (zone.accidentCount >= 4) {
                score += 5;
                assessment.factors.add("Recorded accident history");
            }
        }

        if (isNight()) {
            score += 10;
            assessment.factors.add("Night travel reduces speed judgement");
        }

        assessment.score = Math.min(100, score);
        assessment.level = level(assessment.score);
        assessment.action = action(assessment.level);
        return assessment;
    }

    static int speedKmh(Location location) {
        if (location == null || !location.hasSpeed()) return 0;
        return Math.max(0, Math.round(location.getSpeed() * 3.6f));
    }

    static List<RiskZone> zones() {
        return ZONES;
    }

    private static RiskZone nearestZone(Location location) {
        if (location == null) return null;
        RiskZone best = null;
        double bestDistance = Double.MAX_VALUE;

        for (RiskZone zone : ZONES) {
            double distance = distanceMeters(location, zone);
            if (distance < bestDistance) {
                best = zone;
                bestDistance = distance;
            }
        }

        return best;
    }

    private static double distanceMeters(Location location, RiskZone zone) {
        if (location == null || zone == null) return Double.NaN;

        float[] result = new float[1];
        Location.distanceBetween(
                location.getLatitude(),
                location.getLongitude(),
                zone.latitude,
                zone.longitude,
                result
        );
        return result[0];
    }

    private static boolean isNight() {
        int hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY);
        return hour >= 18 || hour < 6;
    }

    private static String level(int score) {
        if (score >= 81) return "CRITICAL";
        if (score >= 61) return "HIGH";
        if (score >= 31) return "MEDIUM";
        return "LOW";
    }

    private static String action(String level) {
        if ("CRITICAL".equals(level)) return "STOP. Do not enter the junction.";
        if ("HIGH".equals(level)) return "Reduce speed and prepare to stop.";
        if ("MEDIUM".equals(level)) return "Slow down and check all directions.";
        return "Continue with normal caution.";
    }
}

File: public/risk-engine.js
(function attachRiskEngine(global) {
  const LEVELS = [
    { name: "LOW", min: 0, max: 30 },
    { name: "MEDIUM", min: 31, max: 60 },
    { name: "HIGH", min: 61, max: 80 },
    { name: "CRITICAL", min: 81, max: 100 }
  ];

  function clamp(value, min, max) {
    return Math.max(min, Math.min(max, value));
  }

  function kmhToMps(speed) {
    return speed / 3.6;
  }

  function timeToConflictSeconds(distanceMeters, userSpeedKmh, otherSpeedKmh) {
    const closingSpeed = kmhToMps(Math.max(0, userSpeedKmh) + Math.max(0, otherSpeedKmh));
    if (closingSpeed <= 0) return Infinity;
    return distanceMeters / closingSpeed;
  }

  function levelFromScore(score) {
    const normalized = clamp(Math.round(score), 0, 100);
    return LEVELS.find((level) => normalized >= level.min && normalized <= level.max).name;
  }

  function scoreFactor(label, points, evidence) {
    return { label, points: clamp(points, 0, 100), evidence };
  }

  function calculateRisk(input) {
    const scenario = {
      roadType: "Four-road cross junction",
      zoneRisk: "HIGH",
      visibility: "poor",
      timeOfDay: "night",
      userSpeedKmh: 35,
      approachingVehicles: [],
      distanceToConflictMeters: 80,
      accidentCount: 0,
      ...input
    };

    const factors = [];
    const zoneRiskPoints = { LOW: 6, MEDIUM: 14, HIGH: 24, CRITICAL: 30 }[scenario.zoneRisk] || 10;
    factors.push(scoreFactor("Known risk zone", zoneRiskPoints, scenario.zoneRisk));

    if (/junction|crossing|intersection/i.test(scenario.roadType)) {
      factors.push(scoreFactor("Cross-junction conflict points", 18, scenario.roadType));
    }

    if (scenario.timeOfDay === "night") {
      factors.push(scoreFactor("Night travel", 9, "Reduced speed and distance judgement"));
    }

    if (scenario.visibility === "poor") {
      factors.push(scoreFactor("Poor visibility", 14, "Headlights and low lighting reduce judgement"));
    } else if (scenario.visibility === "moderate") {
      factors.push(scoreFactor("Moderate visibility", 7, "Visibility is partially limited"));
    }

    if (scenario.userSpeedKmh > 50) {
      factors.push(scoreFactor("High user speed", 14, `${scenario.userSpeedKmh} km/h`));
    } else if (scenario.userSpeedKmh > 30) {
      factors.push(scoreFactor("Elevated user speed", 8, `${scenario.userSpeedKmh} km/h`));
    }

    if (scenario.accidentCount >= 10) {
      factors.push(scoreFactor("Accident history", 10, `${scenario.accidentCount} recorded incidents`));
    } else if (scenario.accidentCount >= 4) {
      factors.push(scoreFactor("Accident history", 5, `${scenario.accidentCount} recorded incidents`));
    }

    const vehicleFindings = scenario.approachingVehicles.map((vehicle) => {
      const ttc = timeToConflictSeconds(
        Number(vehicle.distanceMeters || scenario.distanceToConflictMeters),
        scenario.userSpeedKmh,
        Number(vehicle.speedKmh || 0)
      );
      const closingSpeedKmh = scenario.userSpeedKmh + Number(vehicle.speedKmh || 0);
      return { ...vehicle, ttc, closingSpeedKmh };
    });

    const mostUrgent = vehicleFindings.reduce((best, current) => {
      if (!best || current.ttc < best.ttc) return current;
      return best;
    }, null);

    if (mostUrgent) {
      const ttc = mostUrgent.ttc;
      let points = 8;
      if (ttc <= 4) points = 30;
      else if (ttc <= 7) points = 24;
      else if (ttc <= 10) points = 16;

      factors.push(
        scoreFactor(
          "Unsafe crossing gap",
          points,
          `${mostUrgent.direction} vehicle, ${Math.round(mostUrgent.closingSpeedKmh)} km/h closing speed, ${ttc.toFixed(1)}s to conflict`
        )
      );

      if (mostUrgent.speedKmh >= 80) {
        factors.push(scoreFactor("Fast approaching vehicle", 12, `${mostUrgent.speedKmh} km/h from ${mostUrgent.direction}`));
      }
    }

    const score = clamp(factors.reduce((sum, factor) => sum + factor.points, 0), 0, 100);
    const level = levelFromScore(score);
    const action =
      level === "CRITICAL"
        ? "STOP. Do not enter the junction."
        : level === "HIGH"
          ? "Reduce speed and prepare to stop."
          : level === "MEDIUM"
            ? "Slow down and check all directions."
            : "Continue with normal caution.";

    return {
      score: Math.round(score),
      level,
      action,
      factors,
      mostUrgent,
      vehicleFindings
    };
  }

  const api = { calculateRisk, timeToConflictSeconds, levelFromScore, kmhToMps };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = api;
  } else {
    global.RiskEngine = api;
  }
})(typeof window !== "undefined" ? window : globalThis);
